Store images in asset store

LourensVeen opened the pull request:
This gives the AssetStore the ability to store Docker images, which I think covers task 1347.

I had a look at using the standard nginx Docker image as a base for the data images, but couldn't get it to work quickly. The current setup is also not exactly exotic, just Ubuntu + nginx with WebDAV enabled, which is a configuration I'm familiar with, so I decided to go with that for now. The whole point of a Docker container is of course that it encapsulates the stuff inside, so this can always be changed later if we're so inclined.

LourensVeen committed:
Add image_location attribute to Asset

LourensVeen committed:
Store containers in asset store

LourensVeen commented:
Converted back to draft, I think I can make this more reviewable...

LourensVeen commented:
I tried to make it a bit smaller, it's still a fair amount of code but I think it could work. I guess a single image would be enough just to test storing images, but I have the rest now so we might as well I guess. Feel free to have a go at reviewing :).

bakhshir commented on proof_of_concept/components/asset_store.py:
```diff
@@ -14,17 +18,29 @@
+        """Releases resources, call when done."""
+        rmtree(self._image_dir, ignore_errors=True)
+
+    def store(self, asset: Asset, move_image: bool = False) -> None:
```
AssetStore.store sounds a bit funny to me, but i can't suggest alternative ('put' does not sound good)

bakhshir commented on proof_of_concept/components/asset_store.py:
```diff
@@ -33,10 +49,17 @@ def store(self, asset: Asset) -> None:
             raise KeyError(f'There is already an asset with id {id}')
 
-        self._assets[asset.id] = asset
+        self._assets[asset.id] = copy(asset)
```
looking back and forth in the library (because i don't know it), this seems ok

LourensVeen replied:
The idea here is that because the asset image is moved or copied, I have to update the Asset object to point to the new location. I don't want to mess up the Asset object passed by the caller, so I'm copying it before modifying. Previously that wasn't necessary, because the Asset wasn't altered.

bakhshir commented on proof_of_concept/components/asset_store.py:
```diff
@@ -14,17 +18,29 @@
+        """Releases resources, call when done."""
+        rmtree(self._image_dir, ignore_errors=True)
+
+    def store(self, asset: Asset, move_image: bool = False) -> None:
```
AssetStore.store sounds a bit funny, but i could not suggest any better alternative ('put' is also meh).

LourensVeen replied:
Hmm, yes, English likes to verb its nouns... Maybe I could rename AssetStore to AssetWarehouse, AssetDepot, AssetBank, AssetDatabase or AssetMuseum (for curated data only!)? Although I guess for AssetDepot I'd have to go with AssetDepot.deposit(), so that may not solve it either :-).

bakhshir commented on proof_of_concept/definitions/identifier.py:
```diff
@@ -44,7 +44,7 @@ def __new__(cls: Type['Identifier'], seq: Any) -> 'Identifier':
                     raise ValueError(f'Invalid identifier segment {segment}')
 
-        return str.__new__(cls, seq)        # type: ignore
+        return str.__new__(cls, seq)
```
does not need to be ignored? ;-)

LourensVeen replied:
Ah yes, welcome to mypy. This is the correct way to create a class derived from str (I think, it's not very extensively documented), but the problem is that this statement returns an object of type `str`, while the signature wants an `Identifier`, so mypy 0.760 gave an error. I'm not sure actually why I used `# type ignore` instead of a cast, but that disabled the error.

Recently, mypy 0.800 was released, and apparently it now knows about deriving from `str` making the `# type: ignore` superfluous. And since superfluous `# type ignore` statements are an error in mypy, I removed it so the tests would pass...

bakhshir reviewed (commented):
Adding docker images handling seems fine to me, I'm not familiar with the dockerization and tests related to it, so I leave it to Sven to review. Otherwise, let me know, I can make an effort.

svenvanderburg commented on tests/test_containers.py:
```diff
@@ -0,0 +1,158 @@
+    return docker.from_env()
+
+
+@pytest.fixture
+def docker_dir(test_dir):
+    """Path to the tests/docker directory."""
+    return test_dir / 'docker'
```
I usually use constants for filepaths such as this. IMO it is more explicit and less verbose than fixtures. What is the advantage of using fixtures for this?

LourensVeen replied:
Mainly symmetry with pytest's `tmpdir` fixture I guess, and my general distaste for globals :-)

svenvanderburg reviewed (approved):
Nice progress! Using ubuntu + nginx + webdav sounds sane.

The changes to `assets.py` and `asset_store.py` are nice and clean. The 'testing' code is a bit less clean. I am not sure how permanent those changes are:
If sort-of permanent :the docker files could benefit from a short documentation comment and `tests/docker/app.py` could maybe be cleaner and clearer with some functions factored out and a bit more documentation.

LourensVeen commented:
I've dockumented the docker files and while I was at it some of the stuff that goes in them, and I've refactored `app.py` into a `libmahiru.py` and `app.py` and cleaned it up a bit. Also, the new `libmahiru.py` now has a global constant _and_ a global variable! Better? :-)

LourensVeen commented:
@bakhshir Can I ask you to have a look at the new `test_asset_store.py`? I decided to remove the Docker-based tests (they'll be back later when I implement execution) and just have a unit test here, as that seems more appropriate.

bakhshir commented on tests/test_asset_store.py:
```diff
@@ -0,0 +1,52 @@
+from pathlib import Path
+from unittest.mock import MagicMock
```
I was reading about MagicMock on friday :)

LourensVeen replied:
It's aptly named, it really is magic. I'm still not entirely sure I understand it, but i can get it to work most of the time, and it's super handy...

bakhshir replied:
I was reading on Friday Mock vs MagicMock. Very interesting, but indeed needs some practice to really get the feeling for it. I get to know quite a bit from the pr reviews, i like it! ;-)

bakhshir reviewed (approved):
test_asset_store.py  looks good, and definitely better readable.

bakhshir reviewed (approved):
sorry, pressed the wrong button :-)
